db/bigtable.go

```diff
@@ -1599,386 +1850,243 @@ func (bigtable *Bigtable) SaveValidatorIncomeDetails(epoch uint64, rewards map[u
+					resultContainer.res[validator] = &itypes.ValidatorEpochIncome{}
 				}
-				resultContainer.mu.Unlock()
-				return true
-			}, gcp_bigtable.RowFilter(filter))
-			return err
-		})
-	}
 
-	if err := g.Wait(); err != nil {
-		return nil, err
+				resultContainer.res[validator].AttestationHeadReward += rewardDetails.AttestationHeadReward
+				resultContainer.res[validator].AttestationSourceReward += rewardDetails.AttestationSourceReward
+				resultContainer.res[validator].AttestationSourcePenalty += rewardDetails.AttestationSourcePenalty
+				resultContainer.res[validator].AttestationTargetReward += rewardDetails.AttestationTargetReward
+				resultContainer.res[validator].AttestationTargetPenalty += rewardDetails.AttestationTargetPenalty
+				resultContainer.res[validator].FinalityDelayPenalty += rewardDetails.FinalityDelayPenalty
+				resultContainer.res[validator].ProposerSlashingInclusionReward += rewardDetails.ProposerSlashingInclusionReward
+				resultContainer.res[validator].ProposerAttestationInclusionReward += rewardDetails.ProposerAttestationInclusionReward
+				resultContainer.res[validator].ProposerSyncInclusionReward += rewardDetails.ProposerSyncInclusionReward
+				resultContainer.res[validator].SyncCommitteeReward += rewardDetails.SyncCommitteeReward
+				resultContainer.res[validator].SyncCommitteePenalty += rewardDetails.SyncCommitteePenalty
+				resultContainer.res[validator].SlashingReward += rewardDetails.SlashingReward
+				resultContainer.res[validator].SlashingPenalty += rewardDetails.SlashingPenalty
+				resultContainer.res[validator].TxFeeRewardWei = utils.AddBigInts(resultContainer.res[validator].TxFeeRewardWei, rewardDetails.TxFeeRewardWei)
+			}
+		}
+		resultContainer.mu.Unlock()
 	}
 
 	return resultContainer.res, nil
 }
 
 // Deletes all block data from bigtable
 func (bigtable *Bigtable) DeleteEpoch(epoch uint64) error {
+	// TOTO: Implement
+	return fmt.Errorf("NOT IMPLEMENTED")
+}
 
-	// First receive all keys that were written by this block (entities & indices)
-	keys := make([]string, 0, 33)
-	startSlot := epoch * utils.Config.Chain.Config.SlotsPerEpoch
-	endSlot := (epoch+1)*utils.Config.Chain.Config.SlotsPerEpoch - 1
+func (bigtable *Bigtable) getValidatorsEpochRanges(validatorIndices []uint64, prefix string, startEpoch uint64, endEpoch uint64) gcp_bigtable.RowRangeList {
 
-	logger.Infof("deleting epoch %v (slot %v to %v)", epoch, startSlot, endSlot)
-	for slot := startSlot; slot <= endSlot; slot++ {
-		keys = append(keys, fmt.Sprintf("%s:e:%s:s:%s", bigtable.chainId, reversedPaddedEpoch(slot/utils.Config.Chain.Config.SlotsPerEpoch), reversedPaddedSlot(slot)))
+	if endEpoch < startEpoch { // handle overflows
+		startEpoch = 0
 	}
-	keys = append(keys, fmt.Sprintf("%s:e:b:%s", bigtable.chainId, reversedPaddedEpoch(epoch)))
 
-	// for _, k := range keys {
-	// 	logger.Info(k)
-	// }
+	ranges := make(gcp_bigtable.RowRangeList, 0, int((endEpoch-startEpoch))*len(validatorIndices)+1)
```

Comment:
**Minor:** Should be
```
ranges := make(gcp_bigtable.RowRangeList, 0, int(endEpoch-startEpoch+1)*len(validatorIndices))
```
Also applies to https://github.com/gobitfly/eth2-beaconchain-explorer/blob/150bfa89b67c1dab30277f30eff6199932542075/db/bigtable.go#L2041